app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java

```diff
@@ -493,15 +534,72 @@ private void queueDownload(Download download) {
         final DownloadManager.Request request = new DownloadManager.Request(Uri.parse(download.getUrl()))
                 .addRequestHeader("User-Agent", download.getUserAgent())
                 .addRequestHeader("Cookie", cookie)
-                .addRequestHeader("Referer", getUrl())
+                .addRequestHeader("Referer", download.getUrl())
                 .setDestinationInExternalPublicDir(download.getDestinationDirectory(), fileName)
                 .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
                 .setMimeType(download.getMimeType());
 
         request.allowScanningByMediaScanner();
 
-        final DownloadManager manager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
-        manager.enqueue(request);
+        manager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
+
+        final IntentFilter filter = new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE);
+
+        getActivity().registerReceiver(downloadBroadcastReceiver, filter);
+
+        mDownloadReference = manager.enqueue(request);
+    }
+
+    public void displaySnackBar(long referenceId) {
```

Comment:
Could we move all this snackbar code into it's own receiver class file? I assume it's pretty independent from this fragment except that it needs to get a reference to a view to show the snackbar.

So the rough idea would be to create an instance of this in onCreateView (That's where we have a view) - then register in onResume() and unregister in onPause() (See below).